pkg/providers/tinkerbell/upgrade.go

```diff
@@ -209,6 +207,45 @@ func (p *Provider) RunPostControlPlaneUpgrade(ctx context.Context, oldClusterSpe
 	return nil
 }
 
+func (p *Provider) ValidateNewSpec(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec) error {
+	prevSpec, err := p.providerKubectlClient.GetEksaCluster(ctx, cluster, clusterSpec.Cluster.Name)
+	if err != nil {
+		return err
+	}
+
+	prevDatacenterConfig, err := p.providerKubectlClient.GetEksaTinkerbellDatacenterConfig(ctx, prevSpec.Spec.DatacenterRef.Name, cluster.KubeconfigFile, prevSpec.Namespace)
+	if err != nil {
+		return err
+	}
+
+	datacenterConfig := p.datacenterConfig
+
+	oSpec := prevDatacenterConfig.Spec
+	nSpec := datacenterConfig.Spec
+
+	prevMachineConfigRefs := machineRefSliceToMap(prevSpec.MachineConfigRefs())
+
+	for _, machineConfigRef := range clusterSpec.Cluster.MachineConfigRefs() {
+		machineConfig, ok := p.machineConfigs[machineConfigRef.Name]
+		if !ok {
+			return fmt.Errorf("cannot find machine config %s in tinkerbell provider machine configs", machineConfigRef.Name)
+		}
+
+		if _, ok = prevMachineConfigRefs[machineConfig.Name]; ok {
+			err = p.validateMachineConfigImmutability(ctx, cluster, machineConfig, clusterSpec)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if nSpec.TinkerbellIP != oSpec.TinkerbellIP {
```

Comment:
Can we also check the optional `OSImageURL` and `HookImagesURLPath`? Those should be changed as well.

Reply:
> OSImageURL will be and must be changed for rolling upgrades. 

Changing these by themselves would not trigger a rollout. Thinking about the cases where these are changed in conjunction with the bare metal upgrade scenarios of k8s version upgrades or scale ups/downs:
- changing hookImageURL should not matter during rolling upgrades, beneficial to allow that even
- changing osImageURL needs to happen during rolling upgrades (validations for this have been discussed before and need to be dealt with separately)
- for scale up/down specifically, both these fields need to be plugged